apollo-router/src/plugins/telemetry/apollo_exporter.rs

```diff
@@ -242,6 +269,22 @@ impl ApolloExporter {
                     } else if status.is_server_error() {
                         tracing::warn!("attempt: {}, could not transfer: {}", i + 1, data);
                         msg = data;
+                        if status == StatusCode::TOO_MANY_REQUESTS {
+                            // We should have a Retry-After header to go with the status code
+                            // If we don't have the header, or it isn't a valid string or we can't
+                            // convert it to u64, just ignore it. Otherwise, interpret it as a
+                            // number of seconds for which we should not attempt to send any more
+                            // reports.
+                            if let Some(retry_after) =
+                                opt_header_retry.and_then(|v| v.to_str().ok()?.parse::<u64>().ok())
+                            {
+                                *STUDIO_BACKOFF.lock().unwrap() =
+                                    Instant::now() + Duration::from_secs(retry_after);
```

Comment:
Don't we want to return a `ApolloExportError::StudioBackoff` at this point?
Also if we do we can remove the mutex as the calling function will know that it must not try to send again until after the backoff expires.

Reply:
Regarding the error, yes I think we should. I don't think I can remove the mutex though since I need it for interior mutability. I could make it a RwLock, but I don't think contention is so high.

Reply:
It's not the contention, it's more that at some point we'll probably want a unit test, and the mutex will make it difficult.
If the timeout is returned can the start method not just have an extra gate? something like:
```
     _ = timeout.tick() && studo_timeout > instant.now()=> {
```
(Not actual code, and there would have to be something for when there is no timeout etc)

Reply:
The other issue is that at the point where we return (i.e.: line 281) I no longer have the original report to return. It is converted into a different report type at line 199... :(

Your suggested change wouldn't cover the cases where traces are reported and the abstraction starts to become very leaky at that point.

I don't think it's the `Mutex` that makes it hard to test but the fact that it's a global. I could make it more testable by passing the global as a parameter, but I don't think I can easily change the "shape" of the `submit_report` function since it's constrained by the opentelemetry API.

Reply:
My latest change replaces the global with a struct member to tidy things up a bit for future testing requirements.

Reply:
I've modified the code to preserve the report. It involves a lot of cloning but does not lose the metrics.